### DISCUSSION TITLE
Welcome to theme-test Discussions!

### DISCUSSION THREAD

**Discussion by A3Bagged** [A3Bagged/theme-test#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by r266-tech** [comment]
Great question! This is **intentional design behavior**, not a bug.

## Why flattening happens

When `add_resource` processes a directory, OpenViking's parser expands each document into a content node, and the storage layer normalizes all nodes into a flat namespace. This is a deliberate tradeoff: flat structures are faster for retrieval (no need to traverse hierarchy during search), and the semantic meaning OpenViking cares about is *document content + metadata*, not folder organization.

## How to preserve hierarchical semantics

If your directory structure carries meaningful categorization, you have a few options:

**Option 1: Use resource namespacing**
Add a custom `namespace` or `tag` when calling `add_resource`, encoding the path hierarchy yourself:
```python
viking.add_resource("papers/nlp/attention.pdf", metadata={"category": "papers/nlp"})
```
Then filter by metadata during retrieval.

**Option 2: Separate resources by top-level directory**
Call `add_resource` separately for each subdirectory with a descriptive name prefix, so the hierarchy is reflected in the resource names.

**Option 3: Encode hierarchy in document metadata**
Before ingestion, preprocess your files to inject directory path into the document metadata fields — OpenViking preserves this through indexing.

The core insight: OpenViking's flat node model trades structural fidelity for retrieval efficiency. If hierarchy is semantically important for your use case, it's best encoded explicitly in metadata rather than relied upon implicitly from folder structure.

Hope this helps!
